File: projects/08/app/src/main/java/io/github/ironking24/nand2tetris/VMTranslator/CodeWriter.java
package io.github.ironking24.nand2tetris.VMTranslator;

import java.io.BufferedWriter;
import java.io.Closeable;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;

import javax.naming.LimitExceededException;

import io.github.ironking24.nand2tetris.VMTranslator.Parser.commandTypes;

/**
 * this class is responsible for translating VM code into assembly instructions and then writing them into the .asm file.
*/
public class CodeWriter implements Closeable{

    private final BufferedWriter writer;
    private String currentFileName, currentFunctionName;
    private final ArrayList<String> statics = new ArrayList<>();
    private final HashMap<String, String> symbolTable = new HashMap<>(9);
    private final HashMap<String, Integer> runningNumbersTable = new HashMap<>(2);

    void xxx(String A) throws IOException{
        writer.write(A);
    }

    CodeWriter(BufferedWriter output) throws IOException{
        this.writer = output;
        this.currentFileName = this.currentFunctionName = "";
        
        //initialize vm to asm translation
        symbolTable.put("add","+");
        symbolTable.put("sub","-");
        symbolTable.put("neg","-");
        symbolTable.put("not","!");
        symbolTable.put("and","&");
        symbolTable.put("or","|");
        symbolTable.put("eq","JEQ");
        symbolTable.put("gt","JGT");
        symbolTable.put("lt","JLT");
        symbolTable.put("argument","ARG");
        symbolTable.put("local","LCL");
        symbolTable.put("this","THIS");
        symbolTable.put("that","THAT");

        //BootStrap
        writer.write(
            "@256\n"+
            "D=A\n"+
            "@SP\n"+
            "M=D\n"
            );
            WriteCall("Sys.init", 0);
    }

    /**
     * sets the file name of the current file being translated.
    */
    public void setFileName(String name){

        currentFileName = name;
    }
    
    /**
     * writes the operation execution on the stack in hack's assembly.
     * @throws IllegalArgumentException if the @param command is not recognized.
    */
    public void writeArithmetic(String command) throws Exception{
        switch(command){
            case "add": case "sub":
            case "and": case "or":
                writer.write(String.format(
                "@SP\n"+
                "AM=M-1\n"+
                "D=M\n"+
                "A=A-1\n"+
                "M=M%sD\n", 
                symbolTable.get(command)));
                break;
            case "neg": case "not":
                writer.write(String.format(
                "@SP\n"+
                "A=M-1\n"+
                "M=%sM\n", 
                symbolTable.get(command)));
                break;
            case "eq": case "gt": case"lt":
                String checkName = String.format("%s$Check%d", currentFunctionName, nextRunningNumber(String.format("%s.%s$Check",currentFileName,currentFunctionName)));
                writer.write(String.format(
                "@SP\n"+
                "AM=M-1\n"+
                "D=M\n"+
                "@SP\n"+
                "AM=M-1\n"+
                "D=M-D\n"+
                "M=-1\n"+
                "@%s\n"+
                "D;%s\n"+
                "@SP\n"+
                "A=M\n"+
                "M=0\n"+
                "(%s)\n"+
                "@SP\n"+
                "AM=M+1\n", 
                checkName, symbolTable.get(command), checkName));
                break;
            default:
                throw new IllegalArgumentException(String.format("Unknown arithmetic command %s.", command));
        }
    }

    /**
     * writes the push/pop execution on the stack in hack's assembly.
     * writes the push/pop execution on the stack in hack's assembly.
     * @throws IllegalArgumentException is thrown when the @param command operation, @param segment or the static is unrecognized.
     * @throws IOException is thrown when I/O error occurs.
     * @throws LimitExceededException is thrown when there is no more room for new statics
     * @throws IndexOutOfBoundsException is thrown when the @param index is exceeds the segment limits
    */
    public void writePushPop(Parser.commandTypes command, String segment, int index) throws IllegalArgumentException, IOException, LimitExceededException{
        switch(command){
            case C_PUSH:
                switch(segment){

                    case "constant":
                        if(index < 0 || index > 32767){
                            throw new IndexOutOfBoundsException(String.format("Illegal constant value: %s", index));
                        }
                        writer.write(String.format(
                        "@%d\n"+
                        "D=A\n",
                        index));
                        break;
                    //local, argument, this, that share the same structure.
                    case "local": case "argument": case "this": case "that":
                        if (index < 0 || index > 32767){
                            throw new IndexOutOfBoundsException(String.format("Illegal segment(%s) index is out of bounds: %s", segment, index));
                        }

                        writer.write(String.format("@%s\n", symbolTable.get(segment)));
                        //index optimization.
                        if(index > 0){
                            writer.write(String.format(
                            "D=M\n"+
                            "@%d\n"+
                            "A=D+A\n",
                             index));
                        }
                        else{
                            writer.write("A=M\n");
                        }
                        writer.write("D=M\n");
                        break;

                    case "pointer":
                        if(!(index >= 0 && index < 2))
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the pointer segment, the index is out of bounds: %d", index));
                        }
                        writer.write(String.format(
                        "@R%d\n"+
                        "D=M\n",
                        index + 3));
                        break;
                    case "temp":
                        if(!(index >= 0 && index <= 7))
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the temp segment, index is out of bounds: %d", index));
                        }
                        writer.write(String.format(
                        "@R%d\n"+
                        "D=M\n",
                        index + 5));
                        break;
                    case "static":
                        if(!statics.contains(String.format("%s.%d", currentFileName, index)))
                        {
                            throw new IllegalArgumentException(String.format("the static address for the push operation (%s.%d) does not exist.", currentFileName, index));
                        }
                        writer.write(String.format(
                        "@%s.%d\n"+
                         "D=M\n"
                        , currentFileName, index));
                        break;
                    default:
                        throw new IllegalArgumentException(String.format("Illegal push segment: %s", segment));
                }
                writer.write(
                 "@SP\n"+
                 "A=M\n"+
                 "M=D\n"+
                 "@SP\n"+
                 "M=M+1\n");
                break;

            case C_POP:
                switch(segment){
                    case "local": case "argument": case "this": case "that":
                        if(index < 0 || index > 32767){
                            throw new IndexOutOfBoundsException(String.format("Illegal segment(%s) index is out of bounds: %s", segment, index));
                        }
                        writer.write(String.format(
                        "@%s\n"+
                        "D=M\n",
                        symbolTable.get(segment)));
                        if(index > 0){
                            writer.write(String.format(
                            "@%d\n"+
                            "D=D+A\n",
                            index));
                        }
                        writer.write(
                        "@R13\n"+
                        "M=D\n"+
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@R13\n"+
                        "A=M\n"+
                        "M=D\n");
                        break;
                    case "pointer":
                        if(index < 0 || index > 2)
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the pointer segment, index is out of bounds: %s", index));
                        }
                        writer.write(String.format(
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@R%d\n"+
                        "M=D\n",
                        index + 3));
                        break;
                    case"temp":
                        if(index < 0 || index > 7)
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the temp segment, index is out of bounds: %s", index));
                        }
                        writer.write(String.format(
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@R%d\n"+
                        "M=D\n",
                        index + 5));
                        break;
                    case "static":
                        if(!statics.contains(String.format("%s.%d", currentFileName, index)))
                        {
                            if(statics.size()+1 > 240){
                                throw new LimitExceededException("The max number of statics has been exceeded.");
                            }
                            statics.add(String.format(
                            "%s.%d",
                            currentFileName, index));
                        }
                        writer.write(String.format(
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@%s.%d\n"+
                        "M=D\n",
                        currentFileName, index));
                        break;
                    default:
                        throw new IllegalArgumentException(String.format("Illegal pop segment: %s", segment));
                }
                break;
            default:
                throw new IllegalArgumentException(String.format("Illegal push/pop command: %s", command));
        }
    }

    /**
     * writes a label assembly.
     * @param  label is whet label name
     * @throws IOException if there is an I/O error.
    */
    public void WriteLabel(String label) throws IOException{
        writer.write(String.format(
            "(%s$%s)\n", 
            currentFunctionName, label));
    }

    /**
     * Writes a goto assembly.
     * @param  label is target of the jump
     * @throws IOException if there is an I/O error.
    */
    public void WriteGoto(String label) throws IOException{
        writer.write(String.format(
        "@%s$%s\n"+
        "0;JMP\n", 
        currentFunctionName, label));
    }

    /**
     * Writes a goto assembly.
     * @param  label is teh jump target if the stack top is true
     * @throws IOException if there is an I/O error.
    */
    public void WriteIf(String label) throws IOException{
        writer.write(String.format(
        "@SP\n"+
        "AM=M-1\n"+
        "D=M\n"+
        "@%s$%s\n"+
        "D;JNE\n",
         currentFunctionName, label));
    }
    
    /**
     * Writes a function assembly.
    */
    public void WriteFunction(String functionName, int nArgs) throws IOException, LimitExceededException{
        currentFunctionName = functionName;
        writer.write(String.format(
            "(%s)\n", 
            functionName));
        //initialize locals to 0
        for(int i = 0; i < nArgs; i++){
            writePushPop(commandTypes.C_PUSH,"constant", 0);
        }
    }

    /**
     * Writes a function call assembly.
     * @param  functionName is the function .
     * @param  nArgs is the number of argument that it has.
     * @throws IOException if there is an I/O error.
    */
    public void WriteCall(String functionName, int nArgs) throws IOException{
        String returnAddress = String.format("%s$ret.%d", functionName, nextRunningNumber(String.format("%s$ret", functionName)));
        writer.write(String.format(
            //push return address
            "@%s\n"+
            "D=A\n"+
            "@SP\n"+
            "A=M\n"+
            "M=D\n"+
            "@SP\n"+
            "M=M+1\n", 
            returnAddress));
            for (int i = 1; i < 5; i++) {
                writer.write(String.format(
                "@%s\n"+
                "D=M\n"+
                "@SP\n"+
                "A=M\n"+
                "M=D\n"+
                "@SP\n"+
                "M=M+1\n",
                "R"+i));
            }
            writer.write(String.format(
            //lcl = sp
            "@SP\n"+
            "D=M\n"+
            "@LCL\n"+
            "M=D\n"+
            //arg = sp-5-nArgs
            "@%d\n"+
            "D=D-A\n"+
            "@ARG\n"+
            "M=D\n"+
            "@%s\n"+
            "0;JMP\n"+
            "(%s)\n",
            nArgs+5, functionName, returnAddress));
    }

    /**
     * Writes a return assembly.
     * @throws LimitExceededException if the static segment if full.
     * @throws IOException if there is an I/O error.
    */
    public void WriteReturn() throws IOException, LimitExceededException {
        //set temp vars
        writer.write(
        //frame (R14) = LCL
        "@LCL\n"+
        "D=M\n"+
        "@R14\n"+
        "M=D\n"+
        //return address = *(frame - 5)
        "@5\n"+
        "A=D-A\n"+
        "D=M\n"+
        "@R15\n"+
        "M=D\n");
        writePushPop(commandTypes.C_POP,"argument", 0);
        writer.write(
        //SP = ARG + 1
        "@ARG\n"+
        "D=M+1\n"+
        "@SP\n"+
        "M=D\n");
        //Ri=*(frame-i)
        for (int i = 4; i > 0; i--) {
            writer.write(String.format(
            "@R14\n"+
            "AM=M-1\n"+
            "D=M\n"+
            "@%s\n"+
            "M=D\n",
            "R"+i));
        }
        writer.write(
        //go to return address
        "@R15\n"+
        "A=M\n"+
        "0;JMP\n");
    }

    /**
     * Closes the BufferedWriter.
     * @throws IOException if there is an I/O error.
    */
    public void close() throws IOException{
        writer.write(
        "(end)\n"+
        "@end\n"+
        "0;JMP\n");
        writer.flush();
        writer.close();
    }

    private int nextRunningNumber(String label){
        if(runningNumbersTable.containsKey(label)){
            return runningNumbersTable.merge(label, 1, Integer::sum);
        }
        else{
            runningNumbersTable.put(label, 0);
            return 0;
        }
    }
}

File: projects/08/src/CodeWriter.cpp
#include "CodeWriter.h"
#include <climits>
#include <string>

namespace VMTranslator 
{
	CodeWriter::CodeWriter(const std::filesystem::path& output_path):output(output_path, std::ios_base::trunc)
	{
		output.exceptions(std::ios_base::badbit | std::ios_base::failbit);

		if (!output.is_open() || !output.good()) 
		{
			throw std::runtime_error("CodeWriter failed to file stream.");
		}

#ifndef WOBS //without bootstrap
		//the define is for testing
		output
			<< "@256\n"
			<< "D=A\n" 
			<< "@SP\n" 
			<< "M=D\n";
		writeCall("Sys.init", 0);
#endif
	}

	void CodeWriter::setFileName(const std::string& name) 
	{
		current_file_name = name;
	}

	void CodeWriter::writeArithmetic(const std::string& command)
	{
		if (command == "add" || command == "sub" || command == "and" || command == "or")
		{
			output
				<< "@SP\n"
				<< "AM=M-1\n"
				<< "D=M\n"
				<< "A=A-1\n"
				<< "M=M" << symbol_table.at(command) << "D\n";
		}
		else if (command == "neg" || command == "not")
		{
			output
				<< "@SP\n"
				<< "A=M-1\n"
				<< "M="<< symbol_table.at(command) << "M\n";
		}
		else if (command == "eq" || command == "gt" || command == "lt")
		{
			std::string checkName;

			if (running_number.find(current_file_name + "$Check") == running_number.cend())
			{
				running_number.emplace(current_file_name + "$Check", 1);
				checkName = current_file_name + "$Check0";
			}
			else 
			{
				int val = running_number[current_file_name + "$Check"]++;
				checkName = current_file_name + "$Check" + std::to_string(val);
			}

			output
				<< "@SP\n"
				<< "AM=M-1\n"
				<< "D=M\n"
				<< "@SP\n"
				<< "AM=M-1\n"
				<< "D=M-D\n"
				<< "M=-1\n"
				<< "@" + checkName + '\n'
				<< "D;" + symbol_table.at(command) + '\n'
				<< "@SP\n"
				<< "A=M\n"
				<< "M=0\n"
				<< "(" + checkName + ")\n"
				<< "@SP\n"
				<< "AM=M+1\n";
		}
	}
	
	void CodeWriter::writePushPop(const Parser::CommandTypes command, const std::string& segment, const short int index)
	{
		switch (command)
		{
		case VMTranslator::Parser::CommandTypes::C_PUSH:
			if (segment == "constant")
			{
				if (index < 0 || index > SHRT_MAX) {
					throw std::runtime_error("Illegal constant value: " + std::to_string(index));
				}

				output
					<< '@' + index + '\n'
					<< "D=A\n";
			}
			else if (segment == "local" || segment == "argument" || segment == "this" || segment == "that")
			{
				if (index < 0 || index > SHRT_MAX) {
					throw std::runtime_error("push " + segment + " index is out of bounds : " + std::to_string(index));
				}
				output << "@" + symbol_table.at(segment) + '\n';
				if (index > 0)
				{
					output
						<< "D=M\n"
						<< "@" + std::to_string(index) + '\n'
						<< "A=D+A\n";
				}
				else
				{
					output << "A=M\n";
				}
				output << "D=M\n";
			}
			else if (segment == "pointer")
			{
				if (index < 0 || index > 1)
				{
					throw std::runtime_error("Illegal push to the pointer segment, the index is out of bounds: " + std::to_string(index));
				}

				output
					<< "@R" + std::to_string(index + 3) + '\n'
					<< "D=M\n";

			}
			else if (segment == "temp")
			{
				if (index < 0 || index > 7)
				{
					throw std::runtime_error("Illegal push to the temp segment, index is out of bounds: " + std::to_string(index));
				}

				output
					<< "@R" << std::to_string(index + 5) + '\n'
					<< "D=M\n";
			}
			else if ("static")
			{
				std::string label = current_file_name + "." + std::to_string(index);

				if (std::find_if(statics.cbegin(), statics.cend(),
					[&](std::string s) {return s == label; }) != statics.cend())
				{
					throw std::runtime_error("the static address does not exist for the push operation of: " + label);
				}
				output
					<< "@" + current_file_name + "." + std::to_string(index) + '\n'
					<< "D=M\n";
			}
			else
			{
				throw std::runtime_error("Illegal push segment: " + segment);
			}
			output
				<< "@SP\n"
				<< "A=M\n"
				<< "M=D\n"
				<< "@SP\n"
				<< "M=M+1\n";
			break;

		case VMTranslator::Parser::CommandTypes::C_POP:
			if (segment == "local" || segment == "argument" || segment == "this" || segment == "that")
			{
				if (index < 0 || index > SHRT_MAX) {
					throw std::runtime_error("pop " + segment + " index is out of bounds : " + std::to_string(index));
				}

				output
					<< "@" + symbol_table.at(segment) + '\n'
					<< "D=M\n";

				if (index > 0) 
				{
					output
						<< "@" + std::to_string(index) + '\n'
						<< "D=D+A\n";
				}
				output
					<< "@R13\n"
					<< "M=D\n"
					<< "@SP\n"
					<< "AM=M-1\n"
					<< "D=M\n"
					<< "@R13\n"
					<< "A=M\n"
					<< "M=D\n";
				break;
			}
			else if (segment == "pointer")
			{
				if (index < 0 || index > 2)
				{
					throw new std::runtime_error("Illegal push to the pointer segment, index is out of bounds: " + std::to_string(index));
				}
				output
					<< "@SP\n"
					<< "AM=M-1\n"
					<< "D=M\n"
					<< "@R" + std::to_string(index + 3) + '\n'
					<< "M=D\n";
			}
			else if (segment == "temp")
			{
				if (index < 0 || index > 7)
				{
					throw std::runtime_error("Illegal push to the temp segment, index is out of bounds: " + std::to_string(index));
				}
				output
					<< "@SP\n"
					<< "AM=M-1\n"
					<< "D=M\n"
					<< "@R" + std::to_string(index + 5) + '\n'
					<< "M=D\n";
			}
			else if ("static")
			{
				if (std::find(statics.cbegin(), statics.cend(), current_file_name + "." + std::to_string(index))!= statics.cend())
				{
					if (statics.size() + 1 > 240) {
						throw std::runtime_error("The max number of statics has been exceeded.");
					}

					statics.push_back(current_file_name + '.' + std::to_string(index));
				}

				output
					<< "@SP\n"
					<< "AM=M-1\n"
					<< "D=M\n"
					<< '@' + current_file_name + '.' + std::to_string(index) + '\n'
					<< "M=D\n";

			}
			else 
			{
				throw std::runtime_error("Illegal pop segment: " + segment);
			}
			break;
		default:
			throw std::runtime_error("Illegal push/pop command at segment: " + segment + " and index " + std::to_string(index));
		}
	}

	void CodeWriter::writeLabel(const std::string& label)
	{
		output
			<< '(' + current_function_name + '$' + label + ')' + '\n';
	}

	void CodeWriter::writeGoto(const std::string& label)
	{
		output 
			<< '@' + current_function_name + '$' + label + '\n'
			<< "0;JMP\n";
	}

	void CodeWriter::writeIf(const std::string& label)
	{
		output
			<< "@SP\n"
			<< "AM=M-1\n"
			<< "D=M\n"
			<< '@' + current_function_name + '$' + label + '\n'
			<< "D;JNE\n";
	}

	void CodeWriter::writeFunction(const std::string& functionName, const int nVars)
	{
		current_function_name = functionName;
		output
			<< '(' + functionName + ')' + '\n';
		//initialize locals to 0
		for (int i = 0; i < nVars; i++) {
			CodeWriter::writePushPop(Parser::CommandTypes::C_PUSH, "constant", 0);
		}
	}

	void CodeWriter::writeCall(const std::string& functionName, const int nVars)
	{

		std::string returnAddress;

		if (running_number.find(functionName + "$ret") == running_number.cend())
		{
			running_number.emplace(current_file_name + "$ret", 1);
			returnAddress = current_file_name + "$ret0";
		}
		else
		{
			int val = running_number[current_file_name + "$ret"]++;
			returnAddress = current_file_name + "$ret" + std::to_string(val);
		}

			//push return address
			output
			<< '@' + returnAddress + '\n'
			<< "D=A\n"
			<< "@SP\n"
			<< "A=M\n"
			<< "M=D\n"
			<< "@SP\n"
			<< "M=M+1\n";

		for (int i = 1; i < 5; i++) {
			output
				<< "@R" + std::to_string(i) + '\n'
				<< "D=M\n"
				<< "@SP\n"
				<< "A=M\n"
				<< "M=D\n"
				<< "@SP\n"
				<< "M=M+1\n";
		}
		output
			//lcl = sp
			<< "@SP\n"
			<< "D=M\n"
			<< "@LCL\n"
			<< "M=D\n"
			//arg = sp-5-nArgs
			<< '@' + std::to_string(nVars + 5) + '\n'
			<< "D=D-A\n"
			<< "@ARG\n"
			<< "M=D\n"
			<< '@' + functionName + '\n'
			<< "0;JMP\n"
			<< '(' + returnAddress + ')' + '\n';
	}

	void CodeWriter::writeReturn()
	{
		//set temp vars
		output
			//frame (R14) = LCL
			<< "@LCL\n"
			<< "D=M\n"
			<< "@R14\n"
			<< "M=D\n" 
			//return address = *(frame - 5)
			<< "@5\n"
			<< "A=D-A\n"
			<< "D=M\n"
			<< "@R15\n"
			<< "M=D\n";
		writePushPop(Parser::CommandTypes::C_POP, "argument", 0);
		output
			//SP = ARG + 1
			<< "@ARG\n"
			<< "D=M+1\n"
			<< "@SP\n"
			<< "M=D\n";
		//Ri=*(frame-i)
		for (int i = 4; i > 0; i--) {
			output
				<< "@R14\n"
				<< "AM=M-1\n"
				<< "D=M\n"
				<< "@R" + std::to_string(i) + '\n'
				<< "M=D\n";
		}
		output
			//go to return address
			<< "@R15\n"
			<< "A=M\n"
			<< "0;JMP\n";
	}

	void CodeWriter::close()
	{
		output
			<< "(end)\n"
			<< "@end\n"
			<< "0;JMP\n";
		output.flush();
		output.close();
	}

	CodeWriter::~CodeWriter()
	{
		if (output.is_open()) 
		{
			output.flush();
			output.close();
		}
	}
}
